walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/interceptors/ExternalInterceptor.kt

```diff
@@ -0,0 +1,73 @@
+package com.microsoft.walletlibrary.did.sdk.datasource.network.interceptors
+
+import com.microsoft.walletlibrary.interceptor.HttpInterceptor
+import com.microsoft.walletlibrary.interceptor.HttpRequest
+import com.microsoft.walletlibrary.interceptor.HttpResponse
+import com.microsoft.walletlibrary.interceptor.MutableHttpHeaders
+import okhttp3.Headers
+import okhttp3.Interceptor
+import okhttp3.Response
+import okhttp3.internal.toImmutableMap
+
+class ExternalInterceptor constructor(private val httpInterceptors: List<HttpInterceptor>) : Interceptor {
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+        // short circuit if no external interceptors
+        if (httpInterceptors.isEmpty()) {
+            return chain.proceed(chain.request())
+        }
+
+        // convert to abstract request
+        val originalRequest = chain.request()
+        val externalRequest = HttpRequest(
+            uri = originalRequest.url.toString(),
+            method = originalRequest.method,
+            headers = this.convertHeaders(originalRequest.headers)
+        )
+
+        // allow each interceptor to mutate the headers
+        httpInterceptors.forEach {
+            it.beforeRequest(externalRequest)
+        }
+        // build a new request if required
+        val newBuilder = originalRequest.newBuilder()
+        var changed = false
```

Comment:
why does changed need to be here?

Reply:
Attempting to reduce rebuilding the request if the interceptor isn't used or doesn't modify headers.